crates/walrus-service/src/node.rs

```diff
@@ -1714,16 +1717,42 @@ impl StorageNodeInner {
         self.node_capability
     }
 
-    pub(crate) fn owned_shards(&self) -> Vec<ShardIndex> {
-        self.committee_service
-            .active_committees()
-            .current_committee()
-            .shards_for_node_public_key(self.public_key())
-            .to_vec()
+    pub(crate) fn owned_shards(&self, epoch: Epoch) -> Vec<ShardIndex> {
+        if self.committee_service.get_epoch() == epoch + 1
+            && self
+                .committee_service
+                .active_committees()
+                .previous_committee()
+                .is_some()
+        {
+            // If epoch is the previous epoch, return the shard assignment based on the previous
+            // committee.
+            self.committee_service
+                .active_committees()
+                .previous_committee()
+                .expect("previous committee should be set")
+                .shards_for_node_public_key(self.public_key())
+                .to_vec()
+        } else {
+            // Otherwise, return the shard assignment based on the current committee.
+            //
+            // Here we don't validate the epoch given that there are already multiple places
+            // in the node logic where storage node cannot be more than 2 epoch lagging. If so,
+            // it'll enter recovery mode.
```

Comment:
Important: I understand the reasoning here, but this could be a bit dangerous when someone just looks at the function name and signature. Could we instead return an error or panic in cases where the event input is not the current or previous epoch? Also, can we describe this behavior in the docstring of the function?

Reply:
I share your concern, and I was particularly wanted to avoid adding a panic. But I do agree that the old function semantics has implicit assumption. What I changed to is to return error if the epoch is neither the current or the previous epoch. The only callsite that will fail is when trying to recover assigned slivers. Currently the node panic and restarting will make the node into recovery mode.